### PR TITLE
config: full delta xDS (including ADS) support

### LINKED ISSUE #4991: Implement incremental xDS in Envoy
The incremental xDS design has been decided on, and added to the appropriate .proto file (in #3470). Now, incremental xDS clients should be implemented in Envoy. I am currently getting started on CDS.

*Relevant Links*: an early draft proposal for incremental xDS, still relevant for examples and overview of goals: https://docs.google.com/document/d/1dh8ACRzNNUkVYB3exepTPlLQK6CyDrOaP2OwwGlxDB4/edit#heading=h.yfahbzhf2783

### PR DESCRIPTION
**Pull request by fredlas**
Delta ADS/xDS support. Added NewGrpcMuxImpl, a GrpcMux implementation that supports delta (and is intended to soon take over for state-of-the-world). DeltaSubscriptionImpl (which can soon be renamed to/replace GrpcSubscriptionImpl) now uses that new class. NewGrpcMuxImpl is the first user of the recently added WatchMap.

Risk Level: medium (huge new thing that people aren't using yet)
Testing: added delta param variant to ads_integration_test
Docs Changes: added source/common/config/README.md and accompanying diagram, added to root/configuration/overview.rst and root/intro/arch_overview/operations/dynamic_configuration.rst.
Release Notes: added support for delta xDS (including ADS) delivery
Fixes #4991

### CODE REVIEW

**Comment by fredlas** [comment]
This PR is not quite as big as it looks. First, most of the 86 changed files are only changed to have the `bool is_delta` plumbing, which is just a temporary measure until NewGrpcMuxImpl unifies implementation of delta+SotW. (Which will be a fairly simple and immediate followup to this PR). Second, WatchMap is not yet merged (#7108).

**Comment by mattklein123** [comment]
Quick drive by: would it be possible to add something about delta vs. SoTW to the arch overview? Maybe here? https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/operations/dynamic_configuration? Maybe also link to the detailed xDS API specification for more details? I think it would be useful to spruce up this page a bit just to talk at a high level about the various options. Thank you!

**Comment by fredlas** [comment]
Yes, great idea, will do!

**Comment by htuch** [comment]
Will review once we have #7108 design level agreement.
/wait

**Comment by alyssawilk** [comment]
At some point can you do a master merge, and move your release note to 1.12.0?  Thanks!

**Comment by fredlas** [comment]
Yup, done.

**Comment by fredlas** [comment]
Reiterating so it doesn't get lost in the shuffle: many of the vast number of files changed only had the "is_delta" param added. That param is an ugly hack to allow this PR to leave the unification of delta+SotW for later. That unification will remove the is_delta from everywhere, and is ready to happen immediately after this PR. (Or even within this PR if you want to do a sort of multi-stage review).

File: docs/root/intro/arch_overview/operations/dynamic_configuration.rst [hunk 1]
```diff
@@ -88,3 +88,13 @@ The :ref:`secret discovery service (SDS) <config_secret_discovery_service>` laye
...
+
+Aggregated xDS ("ADS")
+-----------------------------
+
+EDS, CDS, etc. are each separate services, with different REST/gRPC service names, e.g. StreamListeners, StreamSecrets. For users looking to enforce the order in which resources of different types reach Envoy, there is aggregated xDS, a single gRPC service that carries all resource types in a single gRPC stream. (ADS is only supported by gRPC). :ref:`More details about ADS <config_overview_v2_ads>`.
+
+Delta gRPC xDS
+-----------------------------
+
+Standard xDS is "state-of-the-world": every CDS update must contain every cluster, with the absence of a cluster from an update implying that the cluster is gone. As of 1.11.0, Envoy supports a "delta" variant of xDS (including ADS), where updates only contain resources added/changed/removed. Delta xDS is a new protocol, with request/response protos different from SotW; existing control plane servers will need support added. :ref:`More details about delta <config_overview_v2_delta>`.
```
Comment (lizan) [inline docs/root/intro/arch_overview/operations/dynamic_configuration.rst]: 1.12.0?

File: docs/root/intro/version_history.rst [hunk 2]
```diff
@@ -9,6 +9,7 @@ Version history
 * api: added ::ref:`set_node_on_first_message_only <envoy_api_field_core.ApiConfigSource.set_node_on_first_message_only>` option to omit the node identifier from the subsequent discovery requests on the same stream.
 * config: enforcing that terminal filters (e.g. HttpConnectionManager for L4, router for L7) be the last in their respective filter chains.
 * buffer filter: the buffer filter populates content-length header if not present, behavior can be disabled using the runtime feature `envoy.reloadable_features.buffer_filter_populate_content_length`.
+* config: added support for delta xDS (including ADS) delivery
```
Comment (lizan) [inline docs/root/intro/version_history.rst]: can you link this to the overview doc?

File: include/envoy/config/grpc_mux.h [hunk 3]
```diff
@@ -122,5 +131,35 @@ class GrpcMux {
 
 using GrpcMuxPtr = std::unique_ptr<GrpcMux>;
 
+/**
+ * A grouping of callbacks that a GrpcMux should provide to its GrpcStream.
+ */
+template <class ResponseProto> class GrpcStreamCallbacks {
+public:
+  virtual ~GrpcStreamCallbacks() {}
```
Comment (lizan) [inline include/envoy/config/grpc_mux.h]: nit: = default

File: include/envoy/upstream/cluster_manager.h [hunk 4]
```diff
@@ -180,14 +180,14 @@ class ClusterManager {
...
-   * the management of clusters but instead is required early in ClusterManager/server
+   * Return a reference to the singleton xDS-over-gRPC provider for upstream control plane muxing of
+   * xDS. This is treated somewhat as a special case in ClusterManager, since it does not relate
+   * logically to the management of clusters but instead is required early in ClusterManager/server
    * initialization and in various sites that need ClusterManager for xDS API interfacing.
    *
    * @return GrpcMux& ADS API provider referencee.
    */
-  virtual Config::GrpcMux& adsMux() PURE;
+  virtual std::shared_ptr<Config::GrpcMux> adsMux() PURE;
```
Comment (lizan) [inline include/envoy/upstream/cluster_manager.h]: define GrpcMuxSharedPtr and use it.

File: include/envoy/config/grpc_mux.h [hunk 5]
```diff
@@ -111,6 +113,13 @@ class GrpcMux {
    */
   virtual void resume(const std::string& type_url) PURE;
 
+  // For delta
+  virtual Watch* addOrUpdateWatch(const std::string& type_url, Watch* watch,
```
Comment (lizan) [inline include/envoy/config/grpc_mux.h]: nit: use string_view wherever possible

File: source/common/config/README.md [hunk 6]
```diff
@@ -0,0 +1,47 @@
+# xDS
```
Comment (lizan) [inline source/common/config/README.md]: I'm a bit worried how this file here get attention from developer / updated when things changed. We don't have good structure of developer docs currently though.  @envoyproxy/maintainers thoughts?

File: source/common/config/delta_subscription_impl.cc [hunk 7]
```diff
@@ -1,106 +1,74 @@
...
-      dispatcher_(dispatcher), init_fetch_timeout_(init_fetch_timeout) {}
-
-void DeltaSubscriptionImpl::pause() { state_->pause(); }
-void DeltaSubscriptionImpl::resume() {
-  state_->resume();
-  trySendDiscoveryRequests();
-}
+    std::shared_ptr<GrpcMux> context, absl::string_view type_url, SubscriptionCallbacks& callbacks,
+    SubscriptionStats stats, std::chrono::milliseconds init_fetch_timeout, bool is_aggregated)
+    : context_(context), type_url_(type_url), callbacks_(callbacks), stats_(stats),
```
Comment (lizan) [inline source/common/config/delta_subscription_impl.cc]: context_(std::move(context))

File: source/common/config/delta_subscription_impl.cc [hunk 8]
```diff
@@ -1,106 +1,74 @@
...
-  }
-  return true;
+// Config::SubscriptionCallbacks
+void DeltaSubscriptionImpl::onConfigUpdate(
+    const Protobuf::RepeatedPtrField<ProtobufWkt::Any>& resources,
+    const std::string& version_info) {
+  callbacks_.onConfigUpdate(resources, version_info);
+  stats_.update_success_.inc();
+  stats_.update_attempt_.inc();
+  stats_.version_.set(HashUtil::xxHash64(version_info));
```
Comment (lizan) [inline source/common/config/delta_subscription_impl.cc]: Why do we set the hash to stats_.version_? It doesn't seems informative, calculating xxHash64 from version_info outside of Envoy isn't trivial.

File: source/common/config/delta_subscription_impl.h [hunk 9]
```diff
@@ -1,93 +1,71 @@
...
+  // it's all ours. The practical difference is that we ourselves must call start() on it only in
+  // the latter case.
+  DeltaSubscriptionImpl(std::shared_ptr<GrpcMux> context, absl::string_view type_url,
                         SubscriptionCallbacks& callbacks, SubscriptionStats stats,
-                        std::chrono::milliseconds init_fetch_timeout);
+                        std::chrono::milliseconds init_fetch_timeout, bool is_aggregated);
+  ~DeltaSubscriptionImpl();
 
   void pause();
+
```
Comment (lizan) [inline source/common/config/delta_subscription_impl.h]: nit: unnecessary line

File: source/common/config/new_grpc_mux_impl.h [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+// Manages subscriptions to one or more type of resource. The logical protocol
+// state of those subscription(s) is handled by DeltaSubscriptionState.
+// This class owns the GrpcStream used to talk to the server, maintains queuing
+// logic to properly order the subscription(s)' various messages, and allows
+// starting/stopping/pausing of the subscriptions.
+class NewGrpcMuxImpl : public GrpcMux,
+                       public GrpcStreamCallbacks<envoy::api::v2::DeltaDiscoveryResponse>,
+                       Logger::Loggable<Logger::Id::config> {
+public:
+  NewGrpcMuxImpl(Grpc::RawAsyncClientPtr async_client, Event::Dispatcher& dispatcher,
```
Comment (lizan) [inline source/common/config/new_grpc_mux_impl.h]: nit: RawAsyncClientPtr&&

File: source/common/config/delta_subscription_state.h [hunk 11]
```diff
@@ -86,12 +79,12 @@ class DeltaSubscriptionState : public Logger::Loggable<Logger::Id::config> {
   std::set<std::string> resource_names_;
 
   const std::string type_url_;
+  // callbacks_ is expected to be a WatchMap.
   SubscriptionCallbacks& callbacks_;
   const LocalInfo::LocalInfo& local_info_;
   std::chrono::milliseconds init_fetch_timeout_;
-  Event::TimerPtr init_fetch_timeout_timer_;
+  Event::TimerPtr init_fetch_timeout_timer_{};
```
Comment (lizan) [inline source/common/config/delta_subscription_state.h]: nit: unique_ptr is null initialized by default

File: docs/root/intro/arch_overview/operations/dynamic_configuration.rst (hunk 1 above)
Comment (fredlas) [reply docs/root/intro/arch_overview/operations/dynamic_configuration.rst]: Done.

File: docs/root/intro/version_history.rst (hunk 2 above)
Comment (fredlas) [reply docs/root/intro/version_history.rst]: Done.

File: include/envoy/config/grpc_mux.h (hunk 3 above)
Comment (fredlas) [reply include/envoy/config/grpc_mux.h]: Done.

File: include/envoy/upstream/cluster_manager.h (hunk 4 above)
Comment (fredlas) [reply include/envoy/upstream/cluster_manager.h]: Done.

File: include/envoy/config/grpc_mux.h (hunk 5 above)
Comment (fredlas) [reply include/envoy/config/grpc_mux.h]: This gets used as the key in a map, which does not seem to like string_views.

File: source/common/config/delta_subscription_impl.cc (hunk 7 above)
Comment (fredlas) [reply source/common/config/delta_subscription_impl.cc]: Done.

File: source/common/config/delta_subscription_impl.cc (hunk 8 above)
Comment (fredlas) [reply source/common/config/delta_subscription_impl.cc]: Yeah, it is unfortunate that we have to do this. version_info is a string, but the stats can only report numbers.

File: source/common/config/delta_subscription_impl.h (hunk 9 above)
Comment (fredlas) [reply source/common/config/delta_subscription_impl.h]: Done.

File: source/common/config/new_grpc_mux_impl.h (hunk 10 above)
Comment (fredlas) [reply source/common/config/new_grpc_mux_impl.h]: Done.

File: source/common/config/delta_subscription_state.h (hunk 11 above)
Comment (fredlas) [reply source/common/config/delta_subscription_state.h]: Done.

**Comment by fredlas** [comment]
Thanks! Also, reiterating again so it doesn't get lost in the shuffle:

Many of the vast number of files changed only had the "is_delta" param added. That param is an ugly hack to allow this PR to leave the unification of delta+SotW for later. That unification will remove the is_delta from everywhere, and is ready to happen immediately after this PR. (Or even within this PR if you want to do a sort of multi-stage review).

File: source/common/config/grpc_mux_subscription_impl.cc [hunk 12]
```diff
@@ -9,7 +9,7 @@
 namespace Envoy {
 namespace Config {
 
-GrpcMuxSubscriptionImpl::GrpcMuxSubscriptionImpl(GrpcMux& grpc_mux,
+GrpcMuxSubscriptionImpl::GrpcMuxSubscriptionImpl(GrpcMuxSharedPtr grpc_mux,
                                                  SubscriptionCallbacks& callbacks,
                                                  SubscriptionStats stats,
```
Comment (lambdai) [inline source/common/config/grpc_mux_subscription_impl.cc]: Are you sure you need a temporary `SubScriptionStats` here and copy to `stats_`?

File: source/common/config/grpc_subscription_impl.h [hunk 13]
```diff
@@ -21,27 +21,28 @@ class GrpcSubscriptionImpl : public Config::Subscription {
...
   }
 
-  void updateResources(const std::set<std::string>& update_to_these_names) override {
-    grpc_mux_subscription_.updateResources(update_to_these_names);
+  void updateResourceInterest(const std::set<std::string>& update_to_these_names) override {
+    grpc_mux_subscription_.updateResourceInterest(update_to_these_names);
   }
 
-  GrpcMuxImpl& grpcMux() { return grpc_mux_; }
+  std::shared_ptr<Config::GrpcMuxImpl> grpcMux() { return grpc_mux_; }
```
Comment (lambdai) [inline source/common/config/grpc_subscription_impl.h]: the return type should be `const std::shared_ptr<Config::GrpcMuxImpl>&` to avoid copy shared_ptr

File: source/common/config/new_grpc_mux_impl.cc [hunk 14]
```diff
@@ -0,0 +1,238 @@
...
+  do {
+    for (auto& sub : subscriptions_) {
+      all_subscribed[sub.first] = &sub.second->sub_state_;
+    }
+    for (auto& sub : all_subscribed) {
+      if (already_called.insert(sub).second) { // insert succeeded ==> not already called
+        sub.second->handleEstablishmentFailure();
+      }
+    }
+  } while (all_subscribed.size() != subscriptions_.size());
```
Comment (lambdai) [inline source/common/config/new_grpc_mux_impl.cc]: In the first loop: In what scenario that this condition is false? handleEstablishmentFailure() may mutate `subscriptions_`?

File: include/envoy/upstream/cluster_manager.h [hunk 15]
```diff
@@ -180,14 +180,14 @@ class ClusterManager {
   virtual const envoy::api::v2::core::BindConfig& bindConfig() const PURE;
 
   /**
-   * Return a reference to the singleton ADS provider for upstream control plane muxing of xDS. This
-   * is treated somewhat as a special case in ClusterManager, since it does not relate logically to
-   * the management of clusters but instead is required early in ClusterManager/server
+   * Return a reference to the singleton xDS-over-gRPC provider for upstream control plane muxing of
```
[the remark this answers is not in the document]
Comment (fredlas) [reply include/envoy/upstream/cluster_manager.h]: Done.

File: include/envoy/upstream/cluster_manager.h [hunk 16]
```diff
@@ -225,6 +225,8 @@ class ClusterManager {
   virtual Config::SubscriptionFactory& subscriptionFactory() PURE;
 
   virtual std::size_t warmingClusterCount() const PURE;
+
+  virtual bool xdsIsDelta() const PURE;
```
[the remark this answers is not in the document]
Comment (fredlas) [reply include/envoy/upstream/cluster_manager.h]: Haha, yeah, it certainly is awkward. This is part of that whole temporary "is_delta" machinery I described in the comments in this PR. Added a comment about its pending removal.

File: source/common/config/delta_subscription_impl.h [hunk 17]
```diff
@@ -1,93 +1,70 @@
...
+  void onConfigUpdate(const Protobuf::RepeatedPtrField<ProtobufWkt::Any>& resources,
+                      const std::string& version_info) override;
+  void onConfigUpdate(const Protobuf::RepeatedPtrField<envoy::api::v2::Resource>& added_resources,
+                      const Protobuf::RepeatedPtrField<std::string>& removed_resources,
+                      const std::string& system_version_info) override;
+  void onConfigUpdateFailed(ConfigUpdateFailureReason reason, const EnvoyException* e) override;
+  std::string resourceName(const ProtobufWkt::Any& resource) override;
 
-  void onWriteable() override;
+  GrpcMuxSharedPtr getContextForTest() { return context_; }
```
[the remark this answers is not in the document]
Comment (fredlas) [reply source/common/config/delta_subscription_impl.h]: That is true. I tend to prefer these test-only functions for one-off things like this, and friending for more elaborate access of internal state.

File: source/common/config/delta_subscription_state.cc [hunk 18]
```diff
@@ -1,78 +1,44 @@
...
-                      update_to_these_names.end(), std::inserter(cur_removed, cur_removed.begin()));
-
+void DeltaSubscriptionState::updateSubscriptionInterest(const std::set<std::string>& cur_added,
+                                                        const std::set<std::string>& cur_removed) {
   for (const auto& a : cur_added) {
     setResourceWaitingForServer(a);
-    // Removed->added requires us to keep track of it as a "new" addition, since our user may have
-    // forgotten its copy of the resource after instructing us to remove it, and so needs to be
-    // reminded of it.
+    // If interest in a resource is removed-then-added (all before a discovery request
```
[the remark this answers is not in the document]
Comment (fredlas) [reply source/common/config/delta_subscription_state.cc]: Correct, exactly.

File: source/common/config/delta_subscription_state.h [hunk 19]
```diff
@@ -45,7 +37,8 @@ class DeltaSubscriptionState : public Logger::Loggable<Logger::Id::config> {
 
   void handleEstablishmentFailure();
 
-  envoy::api::v2::DeltaDiscoveryRequest getNextRequest();
+  envoy::api::v2::DeltaDiscoveryRequest getNextRequestAckless();
```
[the remark this answers is not in the document]
Comment (fredlas) [reply source/common/config/delta_subscription_state.h]: Done.

File: source/common/config/grpc_mux_subscription_impl.cc (hunk 12 above)
Comment (fredlas) [reply source/common/config/grpc_mux_subscription_impl.cc]: Because that's how the code already was, and because this class is going to be deleted soon (maybe even in this PR if that's how we decide to proceed), I think it's better to leave this alone.

File: source/common/config/grpc_subscription_impl.h (hunk 13 above)
Comment (fredlas) [reply source/common/config/grpc_subscription_impl.h]: I think that's probably right, but I can't immediately be 100% sure it's safe. This class is going to be entirely replaced anyways, so I would say it's safer and simpler to leave this alone.

File: source/common/config/new_grpc_mux_impl.cc (hunk 14 above)
Comment (fredlas) [reply source/common/config/new_grpc_mux_impl.cc]: Yes, described by the comment above.

Comment (lambdai) [reply source/common/config/new_grpc_mux_impl.cc]: By going through handleEstablishmentFailure overrides I understand the case that CDS as well as other xDS could respond ready() as reaction to failure.

Suppose the newly added subscription is a LDS subscription. My follow up question is why it is critical to invoke `handleEstablishmentFailure()`? You mentioned it as `crash`.
To avoid that crash you check size equal since your assumption is by end of handleEstablishmentFailure() return the new LDS subscription must put itself in `subscriptions_`, 
What if the LDS subscription comes a little later than `handleEstablishmentFailure()`? Crash?

Comment (fredlas) [reply source/common/config/new_grpc_mux_impl.cc]: I'm afraid I don't entirely comprehend what you're saying, but let me try to answer what I think you're asking. This mux thing is responsible for passing on news of the failure to establish the channel along to the various xDS objects it serves, including ones that pop up as a result of telling others the news. If the LDS popped up later, it wouldn't matter, because the failure would have already happened before it popped up. (If you are thinking in terms of concurrency, all of this management code is completely single threaded (thankfully!))

Comment (lambdai) [reply source/common/config/new_grpc_mux_impl.cc]: There is an recent istio change that expecting LDS respond with either all the listeners when envoy starts. 
istio assumes either 
 - all the listeners is in the first SUCCESSFUL response or,
 - in case any error, envoy give up the LDS response.
I am trying to understand if this hold true with this PR.

Questions 1 by 1:
- Without the while, why would envoy crash, according to your comment, when another listener subscription happens after `onEstablishmentFailure`?
- Is the missing notification critical? Since there is retry timer to reconnect and envoy would eventually establish successfully or invoke another `onEstablishmentFailure`

Comment (fredlas) [reply source/common/config/new_grpc_mux_impl.cc]: This part of the code you're looking at is in the context of the xDS gRPC stream failing to even connect. So, there is not any LDS response at all.

Comment (lambdai) [reply source/common/config/new_grpc_mux_impl.cc]: > This part of the code you're looking at is in the context of the xDS gRPC stream failing to even connect. So, there is not any LDS response at all.
Good point. There won't be any successful LDS response, so no partial response either. 

> Is the missing notification critical? 
Other people is more aggressive #8152 Establishment failure is hidden to upper layer. So `all_subscribed.size() != subscriptions_.size()` should not even happen.
I am ok if we leave it here for now and delete in the future once we are fully confident

**Comment by fredlas** [comment]
@lizan 
As I've mentioned a couple of times, this PR is currently a bit messy - the is_delta argument all over the place - due to keeping the work needed to unify delta and state-of-the-world, out of this set of changes.

However, that unified state is the end goal, and I think it has a more natural flow. In fact, I'm starting to question whether it was even worth trying to present the "simpler" delta-only version here.

To speed this process up, and avoid having a whole second PR to be done when this one finally merges, I have merged my long-standing rough drafts of the unified stuff (GrpcMuxImpl, SubscriptionState) into the actual code. There were some conflicts that built up over the past few months, but I've got them mostly worked out; it's passing the main part of ads_integration_test. It is over at https://github.com/envoyproxy/envoy/compare/master...fredlas:unified_ADS

You may still find this current PR easier to review; not sure. But, if you do find the new one easier, let me know, and I'll just merge it straight into here.

**Comment by fredlas** [comment]
Oh! Almost forgot: in the unified version, there are some changes made to naming.

* The existing GrpcMuxImpl, GrpcSubscriptionImpl, GrpcMuxSubscriptionImpl have been deleted.
* The DeltaSubscriptionImpl seen in this PR has been renamed to GrpcSubscriptionImpl; it now also supports state-of-the-world (taking over from GrpcSubscriptionImpl and GrpcMuxSubscriptionImpl).
* The NewGrpcMuxImpl seen in this PR has been renamed to GrpcMuxImpl.
* DeltaSubscriptionState has had a semi-abstract "SubscriptionState" base class factored out of it, so that there can also be a SotwSubscriptionState that GrpcMux can use interchangeably.

**Comment by lizan** [comment]
@fredlas I'll review this again later today, can you merge master to make CI green?

**Comment by fredlas** [comment]
Thanks, done.

File: source/common/config/new_grpc_mux_impl.cc [hunk 20]
```diff
@@ -0,0 +1,238 @@
...
+    updateWatch(type_url, watch, resources);
+    return watch;
+  }
+}
+
+void NewGrpcMuxImpl::removeWatch(const std::string& type_url, Watch* watch) {
+  updateWatch(type_url, watch, {});
+  auto entry = subscriptions_.find(type_url);
+  if (entry == subscriptions_.end()) {
+    ENVOY_LOG(error, "removeWatch() called for non-existent subscription {}.", type_url);
```
Comment (lizan) [inline source/common/config/new_grpc_mux_impl.cc]: No test coverage here, how could this happen? Shouldn't be just an ASSERT?

File: source/common/config/new_grpc_mux_impl.cc [hunk 21]
```diff
@@ -0,0 +1,238 @@
...
+  return pausable_ack_queue_.paused(type_url);
+}
+
+void NewGrpcMuxImpl::onDiscoveryResponse(
+    std::unique_ptr<envoy::api::v2::DeltaDiscoveryResponse>&& message) {
+  ENVOY_LOG(debug, "Received DeltaDiscoveryResponse for {} at version {}", message->type_url(),
+            message->system_version_info());
+  auto sub = subscriptions_.find(message->type_url());
+  if (sub == subscriptions_.end()) {
+    ENVOY_LOG(warn,
```
Comment (lizan) [inline source/common/config/new_grpc_mux_impl.cc]: No coverage here either.

File: source/common/config/new_grpc_mux_impl.cc [hunk 22]
```diff
@@ -0,0 +1,238 @@
+#include "common/config/new_grpc_mux_impl.h"
```
Comment (lizan) [inline source/common/config/new_grpc_mux_impl.cc]: this file doesn't have test coverage in many error cases: https://279265-65214191-gh.circle-artifacts.com/0/coverage/coverage/proc/self/cwd/source/common/config/new_grpc_mux_impl.cc.html

File: source/common/config/delta_subscription_impl.cc [hunk 23]
```diff
@@ -1,106 +1,89 @@
...
-void DeltaSubscriptionImpl::kickOffAck(UpdateAck ack) {
-  ack_queue_.push(ack);
-  trySendDiscoveryRequests();
+// Config::SubscriptionCallbacks
+void DeltaSubscriptionImpl::onConfigUpdate(
+    const Protobuf::RepeatedPtrField<ProtobufWkt::Any>& resources,
+    const std::string& version_info) {
+  callbacks_.onConfigUpdate(resources, version_info);
+  stats_.update_success_.inc();
+  stats_.update_attempt_.inc();
```
[the remark this answers is not in the document]
Comment (fredlas) [reply source/common/config/delta_subscription_impl.cc]: Good question. This exact ordering is ultimately carried over from the old code... and for a second I thought you had found a bug in it!

What actually happens is the throw that can happen within that onConfigUpdate will get caught by the caller of this onConfigUpdate, who will then switch over into onConfigUpdateFailed, which has its own update_attempt.inc().

I think it's safe to call this unacceptably subtle behavior. It's not possible to make it as completely clean as I would like given the current logic, but I have at least changed it so that we're no longer relying on a throw->catch->otherFunction to update the stat. However, looking more closely, I actually don't think the current logic gives the most accurate value for this stat, so I left TODOs.

File: source/common/config/new_grpc_mux_impl.cc (hunk 20 above)
Comment (fredlas) [reply source/common/config/new_grpc_mux_impl.cc]: True, it should not happen, and keeping it from never happening only needs the logic within this class to be right. So, I have made this and most other instances in this file into asserts.

File: source/common/config/new_grpc_mux_impl.cc (hunk 21 above)
Comment (fredlas) [reply source/common/config/new_grpc_mux_impl.cc]: Here is the one instance I will leave as not an assert, since it's about the server's behavior.

Added a test, in new_grpc_mux_impl_test (temporary, will be merged back into grpc_mux_impl_test upon unification).

File: source/common/config/new_grpc_mux_impl.cc (hunk 22 above)
Comment (fredlas) [reply source/common/config/new_grpc_mux_impl.cc]: Done. The ratelimit-related ones are handled in grpc_mux_impl_test, so the unification PR should take care of those.

**Comment by fredlas** [comment]
Well, the tests all passed, but clang_tidy had a couple of issues; should be fixed now. Actually, it had many issues, all in server.cc, but IIUC the issues that are pre-existing in parts of the code I'm not modifying won't be blockers.

**Comment by lambdai** [comment]
LGTM!

File: test/integration/integration.h [hunk 24]
```diff
@@ -220,12 +220,14 @@ class BaseIntegrationTest : Logger::Loggable<Logger::Id::testing> {
   // sending/receiving to/from the (imaginary) xDS server. You should almost always use
   // compareDiscoveryRequest() and sendDiscoveryResponse(), but the SotW/delta-specific versions are
   // available if you're writing a SotW/delta-specific test.
+  // TODO(fredlas) expect_node was defaulting false here; the delta+SotW unification work restores
+  // it.
```
Comment (kyessenov) [inline test/integration/integration.h]: @fredlas We changed the xDS protocol _not_ to send node with every request on a stream, as the node metadata is sometimes larger than the response. I think you might want to revert this change and file a TODO to implement node skipping for delta gRPC.

File: test/integration/rtds_integration_test.cc [hunk 25]
```diff
@@ -36,7 +38,7 @@ std::string tdsBootstrapConfig(absl::string_view api_type) {
           grpc_services:
             envoy_grpc:
               cluster_name: rtds_cluster
-          set_node_on_first_message_only: true
+          set_node_on_first_message_only: false
```
Comment (kyessenov) [inline test/integration/rtds_integration_test.cc]: Please revert the test change. We want to default to this true in the next xDS version.

File: test/integration/integration.h (hunk 24 above)
Comment (fredlas) [reply test/integration/integration.h]: See the text of the TODO: I made sure to accommodate it in the unified version.

This PR was ready to merge a month+ before #7876 was opened, so it would make more sense for that one to be reverted, and reapplied after this one is merged. The ideal, of course, would have been becoming aware of that work much earlier, so that I could recommend it be delayed until after these changes. However, I figured trying to get an already-merged PR reverted would just make things even more slow, and the duplicated effort (since #7876 was written on top of soon-to-be-deleted code) was unfortunately already locked in. So, I reimplemented the intention of #7876 in https://github.com/envoyproxy/envoy/compare/master...fredlas:unified_ADS , which rewrites/refactors gRPC xDS to let this PR's delta additions exist cleanly alongside the existing state-of-the-world logic. Since this PR represents a functional but messy initial version, it is not worth incorporating in here.

Comment (kyessenov) [reply test/integration/integration.h]: As long as we respect the flag for non-incremental xDS, I see no problem, since we can always add this flag to incremental version. I can't immediately tell if that's the case. We are already using this optimization in istio, so I am trying to check if there are regressions in this area.

File: test/integration/rtds_integration_test.cc (hunk 25 above)
Comment (fredlas) [reply test/integration/rtds_integration_test.cc]: Sorry, I think I misread what parts you were commenting on - from what you were saying I thought I might have broken something.

I believe the only node-skipping-related changes I made were here, in the integration testing. The old, to-be-deleted GrpcMuxImpl etc, where it is actually implemented (along with the config loading pipeline to enable it), is unchanged, along with the unit testing. So, no need to be alarmed: this PR does not interfere at all with the node-skipping functionality! It only changes the behavior of the integration tests. I don't think filling-or-not a proto field that is ignored anyways should have any negative effect on the efficacy of the integration testing.

Comment (kyessenov) [reply test/integration/rtds_integration_test.cc]: OK, thanks for confirming. SG if there is no change to node skipping in the old code.

**Comment by fredlas** [comment]
Thanks very much! It is great to have it in. I will try to get the delta+SotW unification in swiftly, so we will have nice clean, modular gRPC xDS code, more amenable to future changes and additions.

**Comment by hzxuzhonghu** [comment]
@fredlas Does the api doc need update? 
```
DELTA_GRPC is not yet entirely implemented! Initially, only CDS is available. Do not use for other xDSes. TODO(fredlas) update/remove this warning when appropriate.
```
https://www.envoyproxy.io/docs/envoy/latest/api-v2/api/v2/core/config_source.proto#enum-core-apiconfigsource-apitype

**Comment by fredlas** [comment]
> @fredlas Does the api doc need update?
> 
> ```
> DELTA_GRPC is not yet entirely implemented! Initially, only CDS is available. Do not use for other xDSes. TODO(fredlas) update/remove this warning when appropriate.
> ```
> 
> https://www.envoyproxy.io/docs/envoy/latest/api-v2/api/v2/core/config_source.proto#enum-core-apiconfigsource-apitype

Looks like it, thanks for pointing it out! I'll add that to #8478

**Comment by wbpcode** [comment]
I have read the relevant documentation and code, but I still don't know how the lazy loading and other functions are implemented. I didn't find the code to update the 'watch' at runtime. Or is this just an infrastructure, and specific features such as lazy loading are not implemented?

**Comment by fredlas** [comment]
Correct, lazy loading is for later. That was actually the motivation behind renaming all of this to "delta", to emphasize the aspect it actually includes, while leaving "incremental" to describe the eventual inclusion of lazy loading. (Which will also need to reach way into Envoy, through the various xDS users, all the way to the data plane, and generally be an even bigger change than this one, I think).
